DEPR: pd.read_table

dahlbaek opened the pull request:
`pd.read_table` is deprecated and replaced by `pd.read_csv`.

- [x] closes #21948
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

dahlbaek committed:
DEPR: pd.read_table

- pd.read_table is deprecated and replaced by pd.read_csv.

- add whatsnew note

- change tests to test for warning messages

- change DataFrame.from_csv to use pandas.read_csv instead of
  pandas.read_table

- Change pandas.read_clipboard to use pandas.read_csv instead
  of pandas.read_table

WillAyd commented:
Initial comments:

- Needs test(s)
- Since we are replacing `read_table` with `read_csv` and a sep of `\t`you should reroute the call to `read_table` to reflect this
 - Needs whatsnew note

pep8speaks commented:
Hello @dahlbaek! Thanks for updating the PR.

 - In the file [`pandas/io/parsers.py`](https://github.com/pandas-dev/pandas/blob/4f41cd04a0839fdac118496bb24fa93dc2d8529b/pandas/io/parsers.py), following are the PEP8 issues :

> [Line 614:20](https://github.com/pandas-dev/pandas/blob/4f41cd04a0839fdac118496bb24fa93dc2d8529b/pandas/io/parsers.py#L614): [E711](https://duckduckgo.com/?q=pep8%20E711) comparison to None should be 'if cond is None:'
> [Line 620:80](https://github.com/pandas-dev/pandas/blob/4f41cd04a0839fdac118496bb24fa93dc2d8529b/pandas/io/parsers.py#L620): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

dahlbaek commented:
Added a whatsnew note.

I'm not sure what to do about tests. `pd.read_csv` has been around for as long as I've been using pandas, so I would suspect that tests are already plentiful.

I changed the warning message so that it will reflect whether or not the user is supplying a value to `sep`. The result is now similar to what I proposed in #21948 but without the decorator, as suggested by @gfyoung. Result:

```pycon
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({"a": [1], "b": [2]})
   ...: df.to_csv("test.csv", sep="\t")

In [3]: pd.read_table("test.csv", index_col=0)
/home/dahlbaek/virtualenvs/pandas-dev/bin/ipython:1: FutureWarning: read_table is deprecated, use read_csv with sep='\t' instead.
  #!/home/dahlbaek/virtualenvs/pandas-dev/bin/python
Out[3]: 
   a  b
0  1  2

In [4]: pd.read_table("test.csv", "\t", index_col=0)
/home/dahlbaek/virtualenvs/pandas-dev/bin/ipython:1: FutureWarning: read_table is deprecated, use read_csv instead.
  #!/home/dahlbaek/virtualenvs/pandas-dev/bin/python
Out[4]: 
   a  b
0  1  2

In [5]: pd.read_table("test.csv", sep="\t", index_col=0)
/home/dahlbaek/virtualenvs/pandas-dev/bin/ipython:1: FutureWarning: read_table is deprecated, use read_csv instead.
  #!/home/dahlbaek/virtualenvs/pandas-dev/bin/python
Out[5]: 
   a  b
0  1  2

In [6]: pd.read_table("test.csv", delimiter="\t", index_col=0)
/home/dahlbaek/virtualenvs/pandas-dev/bin/ipython:1: FutureWarning: read_table is deprecated, use read_csv instead.
  #!/home/dahlbaek/virtualenvs/pandas-dev/bin/python
Out[6]: 
   a  b
0  1  2

In [7]: pd.read_csv("test.csv", "\t", index_col=0)
Out[7]: 
   a  b
0  1  2

In [8]: pd.read_csv("test.csv", sep="\t", index_col=0)
Out[8]: 
   a  b
0  1  2

In [9]: pd.read_csv("test.csv", delimiter="\t", index_col=0)
Out[9]: 
   a  b
0  1  2
```

gfyoung commented:
We definitely have tests for `read_table` in the codebase.  Modify them to assert that a warning is produced.

gfyoung commented:
@dahlbaek : Pass in `check_stacklevel=False` in your tests.

dahlbaek commented:
> @dahlbaek : Pass in check_stacklevel=False in your tests.

Perfect, that's what I figured.

dahlbaek commented:
@WillAyd @gfyoung Does this seem reasonable?

dahlbaek commented on pandas/tests/io/conftest.py:
```diff
@@ -17,7 +18,10 @@ def jsonl_file(datapath):
+    with tm.assert_produces_warning(FutureWarning,
+                                    check_stacklevel=False):
+        df = read_table(datapath('io', 'parser', 'data', 'salaries.csv'))
+    return df
```
Maybe this should just be

```python
@pytest.fixture
def salaries_table(datapath):
    """DataFrame with the salaries dataset"""
    return read_csv(datapath('io', 'parser', 'data', 'salaries.csv'), sep="\t")
```

instead?

dahlbaek commented:
I wonder if it would be better to get rid of the function factory `_make_parser_function` altogether and define `read_csv` and `read_table` separately. It would cause some code duplication, but the code would be easier to read. Also, in the future I would like to propose aligning parameter names of `read_csv` and `to_csv`, and perhaps propose changes to the current `sep`/`delimiter` situation (#7662), but there is no reason those changes should affect `read_table`.

WillAyd commented:
Not sure I completely follow your logic there but it doesn't really seem worth the effort to do, especially since you are deprecating `read_table` here

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -319,6 +319,7 @@ Deprecations
 - :meth:`DataFrame.to_stata`, :meth:`read_stata`, :class:`StataReader` and :class:`StataWriter` have deprecated the ``encoding`` argument.  The encoding of a Stata dta file is determined by the file type and cannot be changed (:issue:`21244`).
 - :meth:`MultiIndex.to_hierarchical` is deprecated and will be removed in a future version  (:issue:`21613`)
 - :meth:`Series.ptp` is deprecated. Use ``numpy.ptp`` instead (:issue:`21614`)
+- :func:`pandas.read_table` is deprecated. Use ``pandas.read_csv`` instead (:issue:`21948`)
```
need a ``:func:`` on read_csv

dahlbaek replied:
Fixed

jreback commented on pandas/tests/io/parser/test_unsupported.py:
```diff
@@ -43,23 +45,43 @@ def test_c_engine(self):
 
         # specify C-unsupported options without python-unsupported options
         with tm.assert_produces_warning(parsers.ParserWarning):
+            warnings.simplefilter("ignore", category=FutureWarning)
```
why are you adding these as read_table? is it to support the read_table deprecation itself? if so, then make a new test function and localize the testing there. if not, then why are you doing it?

dahlbaek replied:
I wanted to assert that both a `parsers.ParserWarning` and a `FutureWarning` is being emitted, and couldn't figure out any other way than first ignoring the `parsers.ParserWarning`, then ignoring the `FutureWarning`. Should I change the `read_table` calls to `read_csv` calls instead? Or simply ignore the `FutureWarning`?

jreback replied:
I think you can use a list of the warnings to assert_produces_warning

dahlbaek replied:
I see, you're right. Maybe it is best to change to `read_csv` anyway, since the tests do not seem to be `read_table` specific?

jreback commented on pandas/tests/test_multilevel.py:
```diff
@@ -519,7 +519,8 @@ def test_xs_level_multiple(self):
 x   q   30      3    -0.6662 -0.5243 -0.3580  0.89145  2.5838"""
 
-        df = read_table(StringIO(text), sep=r'\s+', engine='python')
+        with tm.assert_produces_warning(FutureWarning):
```
anywhere non-essentiall, just change to use read_csv

dahlbaek replied:
Did my best to find non-essential places and do this.

jreback commented:
there is also a check for using pd.read_table in the top-level api that I am not sure is affected but pls check (pandas/tests/api/test_api.py)

dahlbaek commented:
> there is also a check for using pd.read_table in the top-level api that I am not sure is affected but pls check (pandas/tests/api/test_api.py)

I think that test checks the pandas namespace, which I suppose should not change until `pandas.read_table` is actually removed completely. So no change required at this point in time, unless I am mistaken.

jreback commented on pandas/io/parsers.py:
```diff
@@ -606,11 +614,24 @@ def parser_f(filepath_or_buffer,
                  float_precision=None):
 
+        # deprecate read_table
+        if name == "read_table":
```
can you add the issue number here

jreback commented on pandas/tests/io/test_common.py:
```diff
@@ -169,6 +175,23 @@ def test_read_fspath_all(self, reader, module, path, datapath):
         result = reader(mypath)
         expected = reader(path)
+
+        if path.endswith('.pickle'):
```
why was this change needed?

dahlbaek replied:
There is no change. These six lines repeat at the bottom of `test_read_fspath_all_read_table` below, `diff` just does not realise that those six lines are the ones that "ought" to be highlighted instead of these six lines.

jreback commented on pandas/tests/io/test_common.py:
```diff
@@ -169,6 +175,23 @@ def test_read_fspath_all(self, reader, module, path, datapath):
+            tm.assert_frame_equal(result, expected)
+
+    def test_read_fspath_all_read_table(self, datapath):
+        pytest.importorskip('os')
```
why would os not be available?

dahlbaek replied:
I would not know… frankly I just went with what was already written in `test_read_fspath_all`, plugging in `tm.assert_produces_warning` where appropriate.

gfyoung replied:
Let's remove that line then.

jreback reviewed (changes requested):
some comments. can you run the test suite and make sure you are catching all deprecation warnings for read_table.

jreback commented:
@gfyoung if you have any comments

gfyoung commented on pandas/core/frame.py:
```diff
@@ -1592,11 +1592,11 @@ def from_csv(cls, path, header=0, sep=',', index_col=0, parse_dates=True,
+        return read_csv(path, header=header, sep=sep,
+                        parse_dates=parse_dates, index_col=index_col,
+                        encoding=encoding, tupleize_cols=tupleize_cols,
+                        infer_datetime_format=infer_datetime_format)
```
As it stands, this is a breaking API change.

@jreback : `0.21.0` deprecations are fair game to remove in `0.24.0`, no?

dahlbaek replied:
As far as I know, the only difference between `read_csv` and `read_table` is the default value of `sep`. Here, `sep` is explicitly set, so I don't see how this changes anything?

gfyoung replied:
You're right: for the vast majority, it's not going to be breaking.  It's only in a corner case situation (e.g. you pass in `None`) that you will see a change.  This is fine then.

gfyoung commented on pandas/io/clipboards.py:
```diff
@@ -9,7 +9,7 @@
 def read_clipboard(sep=r'\s+', **kwargs):  # pragma: no cover
     r"""
-    Read text from clipboard and pass to read_table. See read_table for the
+    Read text from clipboard and pass to read_csv. See read_csv for the
```
Hmm...this is arguably breaking (the `sep` default is different for both functions), but since we explicitly have a default arg for `sep`, this *probably* should be okay (for at least the vast majority of users).

dahlbaek replied:
Same, maybe I am missing something, but I think the only difference between `read_table` and `read_csv` is anyway being overridden here, thus making the change non-breaking.

gfyoung replied:
Same as above.  It should probably be fine.  Don't worry about it.

gfyoung commented on pandas/io/parsers.py:
```diff
@@ -606,11 +614,24 @@ def parser_f(filepath_or_buffer,
+                              "instead.",
+                              FutureWarning, stacklevel=2)
+            if sep is False:
+                sep = default_sep
```
Why do we need this `if` condition here?

dahlbaek replied:
If the user is setting `sep` or `delimiter`, I want to suggest using `read_csv`, but if neither is set, I also want to suggest passing `sep='\t'`. This was the simplest solution I could come up with which handles the case correclt where the user is passing `sep=';'` to `read_table`.

gfyoung replied:
Hmm...I see.  Okay, fair enough.

dahlbaek commented:
> some comments. can you run the test suite and make sure you are catching all deprecation warnings for read_table.

I'm not seeing any `FutureWarning`s from `read_table`, but I am getting the following failed test:

```
(pandas-dev) $ pytest test_errors.py
=============================== test session starts ===============================
platform linux -- Python 3.6.6, pytest-3.6.3, py-1.5.4, pluggy-0.6.0
rootdir: /home/dahlbaek/git/pandas, inifile: setup.cfg
plugins: xdist-1.22.2, forked-0.2, cov-2.5.1
collected 12 items                                                                

test_errors.py ..........F.                                                 [100%]

==================================== FAILURES =====================================
________________________________ test_error_rename ________________________________

    def test_error_rename():
        # see gh-12665
        from pandas.errors import ParserError
        from pandas.io.common import CParserError
    
        try:
            raise CParserError()
        except ParserError:
            pass
    
        try:
            raise ParserError()
        except CParserError:
            pass
    
        with catch_warnings(record=True):
            try:
>               raise ParserError()
E               pandas.errors.ParserError

test_errors.py:52: ParserError
======================= 1 failed, 11 passed in 0.07 seconds =======================
(pandas-dev) $ 
```

Not sure exactly what is happening here…

gfyoung commented:
@dahlbaek : That's an odd test failure.  Seems to be fine on Travis though.  Also, the test logs don't seem to have any deprecation warnings related to `read_table` AFAICT.

jreback commented:
@dahlbaek we don't usually run *from* the test dir itself, rather from the base dir e.t.
``pytest path/to/test``. running from the test dir might not find the fixtures.

jreback commented:
@jorisvandenbossche @TomAugspurger any objections?

jreback commented:
@dahlbaek see my comments above. The tests are not designed to be run from subdirs, nor are they on our CI. pytest can't find things.

jreback commented:
@dahlbaek hmm that looks like a real error

dahlbaek commented:
@jreback I'll try to look into it when I have time and make a separate issue.

dahlbaek commented:
@jreback I re-downloaded the repo and installed a fresh python3.7 environment. I don't know what was causing the problem, but now it is gone.

jreback commented:
@TomAugspurger if you have any issues here

gfyoung commented:
@dahlbaek : This looks good to me, though if you could fix the merge conflicts, that would be good.

dahlbaek commented:
@gfyoung Done.

TomAugspurger committed:
Add sep note to whatsnew

TomAugspurger commented:
Pushed a doc formatting fix.

Thanks @dahlbaek!
